volmdlr/faces.py

```diff
@@ -1824,16 +1835,91 @@ def bsplinecurve3d_to_2d(self, bspline_curve3d):
         points[-1] = volmdlr.Point2D(theta2, z2)
 
         theta_list = [point.x for point in points]
-        theta_discontinuity = angle_discontinuity(theta_list)
+        theta_discontinuity, indexes_theta_discontinuity = angle_discontinuity(theta_list)
 
         if theta3 < theta1 < theta2 and theta_discontinuity:
             points = [p - volmdlr.Point2D(volmdlr.TWO_PI, 0) if p.x > 0 else p for p in points]
+            if theta2 == 0.0:
+                points[-1] = volmdlr.Point2D(-volmdlr.TWO_PI, z2)
         elif theta3 > theta1 > theta2 and theta_discontinuity:
             points = [p + volmdlr.Point2D(volmdlr.TWO_PI, 0) if p.x < 0 else p for p in points]
+            if theta2 == 0.0:
+                points[-1] = volmdlr.Point2D(volmdlr.TWO_PI, z2)
+        elif theta_discontinuity and len(indexes_theta_discontinuity) == 1:
+            index_theta_discontinuity = indexes_theta_discontinuity[0]
+            sign = points[index_theta_discontinuity - 1].x / abs(points[index_theta_discontinuity - 1].x)
+            points = [p + sign * volmdlr.Point2D(volmdlr.TWO_PI, 0) if i >= index_theta_discontinuity else p
+                      for i, p in enumerate(points)]
+        elif theta_discontinuity:
+            raise NotImplementedError
 
         return [vme.BSplineCurve2D.from_points_interpolation(points, degree=bspline_curve3d.degree,
                                                              periodic=bspline_curve3d.periodic)]
 
+    def arcellipse3d_to_2d(self, arcellipse3d):
+        """
+        Transformation of a 3D arc of ellipse to a 2D primitive in a cylindrical surface.
+
+        """
+        points = [self.point3d_to_2d(p)
+                  for p in arcellipse3d.discretization_points(number_points=50)]
+
+        theta1, z1 = self.point3d_to_2d(arcellipse3d.start)
+        theta2, z2 = self.point3d_to_2d(arcellipse3d.end)
+
+        # theta3, _ = self.point3d_to_2d(arcellipse3d.point_at_abscissa(0.001 * length))
+        theta3, _ = points[1]
+        # make sure that the reference angle is not undefined
+        if abs(theta3) == math.pi:
+            theta3, _ = points[1]
+
+        # Verify if theta1 or theta2 point should be -pi because atan2() -> ]-pi, pi]
+        if abs(theta1) == math.pi or abs(theta1) == 0.5 * math.pi:
+            theta1 = vm_parametric.repair_start_end_angle_periodicity(theta1, theta3)
+        if abs(theta2) == math.pi or abs(theta2) == 0.5 * math.pi:
+            theta4, _ = points[-2]
+            # make sure that the reference angle is not undefined
+            if abs(theta4) == math.pi:
+                theta4, _ = points[-3]
+            theta2 = vm_parametric.repair_start_end_angle_periodicity(theta2, theta4)
```

Comment:
should all the researches of different thetas be replaced in another method maybe?